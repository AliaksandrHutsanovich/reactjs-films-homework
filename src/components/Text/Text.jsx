import React, { memo, useMemo } from 'react';
import PropTypes from 'prop-types';
import { Typography } from 'antd';
import clsx from 'clsx';

import '!style-loader!css-loader!sass-loader!antd/dist/antd.css'; //eslint-disable-line

import styles from './Text.scss';

const { Text: AntdText } = Typography;

const classNameTypes = {
  white: styles.white,
  with_shadow: styles.withShadow,
  white_with_background: styles.whiteWithBackground,
  secondary_with_background: styles.secondaryWithBackground,
  secondary: styles.secondary,
  bordered: styles.bordered,
};

const getSizeClass = (size) => styles[`size_${size}`];

const Text = ({
  underline,
  strong,
  type,
  children,
  size,
  textOverflow,
}) => {
  const className = classNameTypes[type];

  const antdTextType = useMemo(() => (
    type === 'warning' || type === 'danger' ? type : ''
  ), [type]);

  const fontSizeClass = useMemo(() => getSizeClass(size), [size]);

  return (
    <AntdText
      className={clsx(
        className,
        fontSizeClass,
        { [styles.bold]: strong },
        { [styles.ellipsis]: textOverflow },
      )}
      underline={underline}
      strong={strong}
      type={antdTextType}
    >
      {children}
    </AntdText>
  );
};

Text.propTypes = {
  underline: PropTypes.bool,
  strong: PropTypes.bool,
  type: PropTypes.oneOf([
    ...Object.keys(classNameTypes),
    'normal',
    'warning',
    'danger',
  ]).isRequired,
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  size: PropTypes.number,
  textOverflow: PropTypes.bool,
};

Text.defaultProps = {
  underline: false,
  strong: false,
  size: 14,
  textOverflow: false,
};

export default memo(Text);
