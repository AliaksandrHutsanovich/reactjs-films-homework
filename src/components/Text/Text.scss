@font-face{
  font-family: 'HelveticaNeue';
  src: url('../../assets/fonts/HelveticaNeue.ttf') format('truetype');
}

%font-shared {
  font-family: HelveticaNeue;
}

.secondary {
  @extend %font-shared;
  color: rgb(71, 195, 240) !important;
}

.white {
  @extend %font-shared;
  color: white;
}

@mixin fillBackground($color) {
  display: block;
  color: white;
  margin: 0;
  line-height: 12px;
  padding: 4px 8px;
  @extend %font-shared;
  background-color: $color;
}

.withShadow {
  @include fillBackground( rgba(81, 79, 60, 0.5));
  padding: 10px 13px;
  line-height: 19px;
}

.whiteWithBackground {
  @include fillBackground( rgb(179, 179, 179));
  border-radius: 2px;
}

.secondaryWithBackground {
  @extend .whiteWithBackground;
  @extend .secondary;
}

@for $var from 0 to 11 {
  .size_#{10 + $var * 2} {
    font-size: #{10 + $var * 2}px;
  }
}

.bold {
  font-weight: bold;
}

.ellipsis {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.bordered {
  @extend .secondary;
  border: {
    radius: 6px;
    width: 1px;
    style: solid;
  }
  padding: 3px 7px;
}
