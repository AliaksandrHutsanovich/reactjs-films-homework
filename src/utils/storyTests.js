const storyTests = (stories, renderer) => {
  Object.keys(stories).forEach((storyKey) => {
    if (typeof stories[storyKey] !== 'function') {
      return;
    }
    it(`${storyKey} should be rendered properly`, () => {
      renderer.render(stories[storyKey]());
      const result = renderer.getRenderOutput();

      expect(result).toMatchSnapshot();
    });
  });
};

export default storyTests;
