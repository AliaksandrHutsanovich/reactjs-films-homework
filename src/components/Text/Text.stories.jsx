import React from 'react';
import Text from './Text';
import image from '../../assets/images/story-background.png';

const backgroundDecorator = (story) => (
  <div
    style={{
      background: `url(${image}) no-repeat`,
      padding: '20px',
      width: '300px',
      height: '300px',
    }}
  >
    {story()}
  </div>
);

export default {
  component: Text,
  title: 'components/Text',
};

export const white = () => (
  <Text type="white">
    This is white text
  </Text>
);
white.story = {
  decorators: [backgroundDecorator],
};

export const withShadowBackground = () => (
  <Text type="with_shadow">
    This is text with shadow background
  </Text>
);
withShadowBackground.story = {
  decorators: [backgroundDecorator],
};

export const withBackground = () => (
  <Text type="white_with_background">
    This is text with background
  </Text>
);
withBackground.story = {
  decorators: [backgroundDecorator],
};

export const secondary = () => (
  <Text type="secondary">
    This is text with secondary color
  </Text>
);

export const normal = () => (
  <Text type="normal">
    This is normal text
  </Text>
);

export const danger = () => (
  <Text type="danger">
    This is normal text
  </Text>
);
