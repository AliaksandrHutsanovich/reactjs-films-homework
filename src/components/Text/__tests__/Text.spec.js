import ShallowRenderer from 'react-test-renderer/shallow';
import storyTests from '../../../utils/storyTests';
import * as stories from '../Text.stories';

describe('<Text />', () => {
  const renderer = new ShallowRenderer();

  storyTests(stories, renderer);
});
