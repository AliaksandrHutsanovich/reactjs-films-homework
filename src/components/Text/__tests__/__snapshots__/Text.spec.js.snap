// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> bordered should be rendered properly 1`] = `
<Text
  className="bordered size_undefined"
  type=""
>
  It is bordered text
</Text>
`;

exports[`<Text /> danger should be rendered properly 1`] = `
<Text
  className="size_undefined"
  type="danger"
>
  This is normal text
</Text>
`;

exports[`<Text /> normal should be rendered properly 1`] = `
<Text
  className="size_undefined"
  type=""
>
  This is normal text
</Text>
`;

exports[`<Text /> secondary should be rendered properly 1`] = `
<Text
  className="secondary size_undefined"
  type=""
>
  This is text with secondary color
</Text>
`;

exports[`<Text /> white should be rendered properly 1`] = `
<Text
  className="white size_undefined"
  type=""
>
  This is white text
</Text>
`;

exports[`<Text /> withBackground should be rendered properly 1`] = `
<Text
  className="whiteWithBackground size_undefined"
  type=""
>
  This is text with background
</Text>
`;

exports[`<Text /> withShadowBackground should be rendered properly 1`] = `
<Text
  className="withShadow size_undefined"
  type=""
>
  This is text with shadow background
</Text>
`;
